lib/src/neu/parsers/type-parser.ts

```diff
@@ -398,6 +402,141 @@ function parseUnionType(
+    typeTable,
+    lociTable
+  );
+  if (refTypeResult.isErr()) return refTypeResult;
+  const refType = refTypeResult.unwrap();
+  if (refType.kind !== TypeKind.REFERENCE) {
+    return err(
+      new TypeNotAllowedError("Indexed access type must be reference", {
+        file: typeNode.getSourceFile().getFilePath(),
+        position: typeNode.getPos()
+      })
+    );
+  }
+  const resolvedType = resolveIndexedAccessType(
+    propertyAccessChainResult.unwrap(),
+    refType,
+    typeTable
+  );
+
+  return ok(resolvedType);
+}
+
+/**
+ * Resolve the target type for an indexed access type.
+ *
+ * @param propertyChain properties to traverse
+ * @param currentType type to inspect
+ * @param typeTable a TypeTable
+ */
+function resolveIndexedAccessType(
+  propertyChain: string[],
+  currentType: Type,
+  typeTable: TypeTable
+): Type {
+  if (propertyChain.length === 0) return currentType;
+  if (currentType.kind === TypeKind.OBJECT) {
+    const property = currentType.properties.find(
+      p => p.name === propertyChain[0]
+    );
+    if (property === undefined) {
+      throw new Error("Indexed type property not found");
+    }
+    return resolveIndexedAccessType(
+      propertyChain.slice(1),
+      property.type,
+      typeTable
+    );
+  }
+  if (currentType.kind === TypeKind.REFERENCE) {
+    const referencedType = typeTable.getOrError(currentType.name);
+    return resolveIndexedAccessType(propertyChain, referencedType, typeTable);
+  }
+  throw new Error("Indexed type error");
+}
+
+/**
+ * Resolve the root reference type of an indexed access type.
+ *
+ * @param typeNode an indexed access type node
+ */
+function resolveIndexedAccessRootReference(
+  typeNode: IndexedAccessTypeNode
+): Result<TypeReferenceNode, ParserError> {
+  const objectType = typeNode.getObjectTypeNode();
+  if (TypeGuards.isIndexedAccessTypeNode(objectType)) {
+    return resolveIndexedAccessRootReference(objectType);
+  }
+  if (!TypeGuards.isTypeReferenceNode(objectType)) {
+    return err(
+      new TypeNotAllowedError("Indexed access type must be reference", {
+        file: objectType.getSourceFile().getFilePath(),
+        position: objectType.getPos()
+      })
+    );
+  }
+  return ok(objectType);
+}
+
+/**
+ * Resolve the property access chain of an indexed access type.
+ *
+ * @param typeNode an indexed access type node
+ * @param accResult property chain result accumulator
+ */
+function resolveIndexAccessPropertyAccessChain(
+  typeNode: IndexedAccessTypeNode,
+  accResult: Result<string[], ParserError> = ok([])
+): Result<string[], ParserError> {
+  if (accResult.isErr()) return accResult;
+  const acc = accResult.unwrap();
+
+  const literalTypeNode = typeNode.getIndexTypeNode();
+  if (!TypeGuards.isLiteralTypeNode(literalTypeNode)) {
+    throw new Error("expected type literal");
+  }
+  const literalTypeResult = parseLiteralType(literalTypeNode);
+  if (literalTypeResult.isErr()) return literalTypeResult;
```

Comment:
```suggestion
  if (literalTypeResult.isErr()) {
   return literalTypeResult;
  }
```
To preserve consistency